File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/shell/Inspector.kt
package xyz.ksharma.prisma.catalogue.shell

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import xyz.ksharma.prisma.components.icons.PrismaIcons
import xyz.ksharma.prisma.coreui.LocalPrismaIsDark
import xyz.ksharma.prisma.coreui.themed
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColor
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography

/**
 * Drives the right-side Inspector panel. Toggleable from chrome; visibility
 * is conversation-scoped (no rememberSaveable) because it's a transient
 * debug aid — coming back from the background should land cleanly closed.
 */
public class InspectorController(
    public val isOpen: Boolean,
    public val toggle: () -> Unit,
    public val close: () -> Unit,
)

public val LocalInspectorController = compositionLocalOf<InspectorController> {
    error("InspectorController not provided. Wrap your tree in CatalogueRoot.")
}

/**
 * Right-side runtime token reference. Lists every semantic colour with its
 * resolved hex for the active theme, plus the spacing / radius / typography
 * scales. A stop-gap for "what's this token actually rendering as right now?"
 * questions during development.
 */
@Composable
public fun InspectorOverlay(
    controller: InspectorController,
    modifier: Modifier = Modifier,
) {
    AnimatedVisibility(
        visible = controller.isOpen,
        enter = fadeIn(),
        exit = fadeOut(),
        modifier = modifier.fillMaxSize(),
    ) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 0.4f))
                .clickable(onClick = controller.close),
        ) {
            AnimatedVisibility(
                visible = controller.isOpen,
                enter = slideInHorizontally { it },
                exit = slideOutHorizontally { it },
                modifier = Modifier.align(Alignment.CenterEnd),
            ) {
                InspectorPanel(onClose = controller.close)
            }
        }
    }
}

@Composable
private fun InspectorPanel(onClose: () -> Unit) {
    Column(
        modifier = Modifier
            .fillMaxHeight()
            .widthIn(min = 320.dp, max = 420.dp)
            .background(PrismaSemanticColors.SurfaceBase.themed())
            .border(
                width = 1.dp,
                color = PrismaSemanticColors.BorderSubtle.themed(),
            )
            .clickable(enabled = false, onClick = {}),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = PrismaSpacing.Sp5, vertical = PrismaSpacing.Sp4),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Text(
                text = "Inspector",
                style = PrismaTypography.TitleMd,
                color = PrismaSemanticColors.TextPrimary.themed(),
            )
            Box(
                modifier = Modifier
                    .size(28.dp)
                    .clip(CircleShape)
                    .clickable(role = Role.Button, onClick = onClose),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    painter = painterResource(PrismaIcons.Close),
                    contentDescription = "Close inspector",
                    tint = PrismaSemanticColors.TextSecondary.themed(),
                    modifier = Modifier.size(16.dp),
                )
            }
        }

        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = PrismaSpacing.Sp5)
                .background(PrismaSemanticColors.BorderSubtle.themed())
                .padding(top = 1.dp),
        )

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = PrismaSpacing.Sp5, vertical = PrismaSpacing.Sp5),
            verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp7),
        ) {
            Group("Surface") {
                ColorRow("SurfaceBase", PrismaSemanticColors.SurfaceBase)
                ColorRow("SurfaceRaised", PrismaSemanticColors.SurfaceRaised)
                ColorRow("SurfaceSunken", PrismaSemanticColors.SurfaceSunken)
                ColorRow("SurfaceInverse", PrismaSemanticColors.SurfaceInverse)
                ColorRow("SurfaceOverlay", PrismaSemanticColors.SurfaceOverlay)
            }
            Group("Text") {
                ColorRow("TextPrimary", PrismaSemanticColors.TextPrimary)
                ColorRow("TextSecondary", PrismaSemanticColors.TextSecondary)
                ColorRow("TextTertiary", PrismaSemanticColors.TextTertiary)
                ColorRow("TextDisabled", PrismaSemanticColors.TextDisabled)
                ColorRow("TextOnAccent", PrismaSemanticColors.TextOnAccent)
                ColorRow("TextLink", PrismaSemanticColors.TextLink)
            }
            Group("Border") {
                ColorRow("BorderSubtle", PrismaSemanticColors.BorderSubtle)
                ColorRow("BorderDefault", PrismaSemanticColors.BorderDefault)
                ColorRow("BorderStrong", PrismaSemanticColors.BorderStrong)
                ColorRow("BorderFocus", PrismaSemanticColors.BorderFocus)
            }
            Group("Accent") {
                ColorRow("AccentDefault", PrismaSemanticColors.AccentDefault)
                ColorRow("AccentHover", PrismaSemanticColors.AccentHover)
                ColorRow("AccentPressed", PrismaSemanticColors.AccentPressed)
                ColorRow("AccentSubtle", PrismaSemanticColors.AccentSubtle)
            }
            Group("Status") {
                ColorRow("Success", PrismaSemanticColors.StatusSuccessDefault)
                ColorRow("Warning", PrismaSemanticColors.StatusWarningDefault)
                ColorRow("Danger", PrismaSemanticColors.StatusDangerDefault)
                ColorRow("Info", PrismaSemanticColors.StatusInfoDefault)
            }
            Group("Spacing") {
                ScalarRow("Sp1", "${PrismaSpacing.Sp1.value.toInt()} dp")
                ScalarRow("Sp2", "${PrismaSpacing.Sp2.value.toInt()} dp")
                ScalarRow("Sp3", "${PrismaSpacing.Sp3.value.toInt()} dp")
                ScalarRow("Sp4", "${PrismaSpacing.Sp4.value.toInt()} dp")
                ScalarRow("Sp5", "${PrismaSpacing.Sp5.value.toInt()} dp")
                ScalarRow("Sp6", "${PrismaSpacing.Sp6.value.toInt()} dp")
                ScalarRow("Sp7", "${PrismaSpacing.Sp7.value.toInt()} dp")
                ScalarRow("Sp8", "${PrismaSpacing.Sp8.value.toInt()} dp")
            }
            Group("Radius") {
                ScalarRow("Sm", "${PrismaRadius.Sm.value.toInt()} dp")
                ScalarRow("Md", "${PrismaRadius.Md.value.toInt()} dp")
                ScalarRow("Lg", "${PrismaRadius.Lg.value.toInt()} dp")
                ScalarRow("Full", "${PrismaRadius.Full.value.toInt()} dp")
            }
            Spacer(modifier = Modifier.size(PrismaSpacing.Sp7))
        }
    }
}

@Composable
private fun Group(label: String, content: @Composable () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp3)) {
        Text(
            text = label.uppercase(),
            style = PrismaTypography.LabelSm,
            color = PrismaSemanticColors.TextTertiary.themed(),
        )
        Column(verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2)) {
            content()
        }
    }
}

@Composable
private fun ColorRow(label: String, token: PrismaSemanticColor) {
    val isDark = LocalPrismaIsDark.current
    val resolved = token.resolve(isDark)
    val hex = "#%02X%02X%02X".format(
        (resolved.red * 255).toInt(),
        (resolved.green * 255).toInt(),
        (resolved.blue * 255).toInt(),
    )
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp3),
    ) {
        Box(
            modifier = Modifier
                .size(20.dp)
                .clip(RoundedCornerShape(4.dp))
                .background(resolved)
                .border(1.dp, PrismaSemanticColors.BorderSubtle.themed(), RoundedCornerShape(4.dp)),
        )
        Text(
            text = label,
            style = PrismaTypography.BodySm,
            color = PrismaSemanticColors.TextPrimary.themed(),
            modifier = Modifier.width(140.dp),
        )
        Text(
            text = hex,
            style = PrismaTypography.BodySm.copy(fontFamily = FontFamily.Monospace),
            color = PrismaSemanticColors.TextTertiary.themed(),
        )
    }
}

@Composable
private fun ScalarRow(label: String, value: String) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp3),
    ) {
        Text(
            text = label,
            style = PrismaTypography.BodySm,
            color = PrismaSemanticColors.TextPrimary.themed(),
            modifier = Modifier.width(80.dp),
        )
        Text(
            text = value,
            style = PrismaTypography.BodySm.copy(fontFamily = FontFamily.Monospace),
            color = PrismaSemanticColors.TextTertiary.themed(),
        )
    }
}

File: ios-catalogue/App/Shell/Inspector.swift
import SwiftUI
import CoreUI
import Components
#if canImport(UIKit)
import UIKit
#endif

/// Right-side runtime token reference. Lists every semantic colour with its
/// resolved hex for the active theme, plus the spacing / radius scales.
/// A stop-gap for "what's this token actually rendering as right now?"
/// questions during development.
struct InspectorOverlay: View {
    @Environment(\.colorScheme) private var scheme
    @Binding var open: Bool

    var body: some View {
        ZStack {
            if open {
                Color.black.opacity(0.4)
                    .ignoresSafeArea()
                    .transition(.opacity)
                    .onTapGesture { open = false }

                HStack(spacing: 0) {
                    Spacer()
                    InspectorPanel(onClose: { open = false })
                        .frame(maxWidth: 420)
                        .frame(maxHeight: .infinity)
                        .transition(.move(edge: .trailing))
                }
                .ignoresSafeArea()
            }
        }
        .animation(.easeInOut(duration: 0.25), value: open)
    }
}

private struct InspectorPanel: View {
    @Environment(\.colorScheme) private var scheme
    let onClose: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack {
                Text("Inspector")
                    .font(PrismaTypography.titleMd.font)
                    .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
                Spacer()
                Button(action: onClose) {
                    Image(prisma: .close)
                        .renderingMode(.template)
                        .resizable()
                        .frame(width: 16, height: 16)
                        .foregroundStyle(PrismaSemanticColors.textSecondary.themed(scheme))
                        .frame(width: 28, height: 28)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Close inspector")
            }
            .padding(.horizontal, PrismaSpacing.sp5)
            .padding(.vertical, PrismaSpacing.sp4)

            Divider().background(PrismaSemanticColors.borderSubtle.themed(scheme))

            ScrollView {
                VStack(alignment: .leading, spacing: PrismaSpacing.sp7) {
                    group("Surface") {
                        colorRow("surfaceBase", PrismaSemanticColors.surfaceBase)
                        colorRow("surfaceRaised", PrismaSemanticColors.surfaceRaised)
                        colorRow("surfaceSunken", PrismaSemanticColors.surfaceSunken)
                        colorRow("surfaceInverse", PrismaSemanticColors.surfaceInverse)
                        colorRow("surfaceOverlay", PrismaSemanticColors.surfaceOverlay)
                    }
                    group("Text") {
                        colorRow("textPrimary", PrismaSemanticColors.textPrimary)
                        colorRow("textSecondary", PrismaSemanticColors.textSecondary)
                        colorRow("textTertiary", PrismaSemanticColors.textTertiary)
                        colorRow("textDisabled", PrismaSemanticColors.textDisabled)
                        colorRow("textOnAccent", PrismaSemanticColors.textOnAccent)
                        colorRow("textLink", PrismaSemanticColors.textLink)
                    }
                    group("Border") {
                        colorRow("borderSubtle", PrismaSemanticColors.borderSubtle)
                        colorRow("borderDefault", PrismaSemanticColors.borderDefault)
                        colorRow("borderStrong", PrismaSemanticColors.borderStrong)
                        colorRow("borderFocus", PrismaSemanticColors.borderFocus)
                    }
                    group("Accent") {
                        colorRow("accentDefault", PrismaSemanticColors.accentDefault)
                        colorRow("accentHover", PrismaSemanticColors.accentHover)
                        colorRow("accentPressed", PrismaSemanticColors.accentPressed)
                        colorRow("accentSubtle", PrismaSemanticColors.accentSubtle)
                    }
                    group("Status") {
                        colorRow("Success", PrismaSemanticColors.statusSuccessDefault)
                        colorRow("Warning", PrismaSemanticColors.statusWarningDefault)
                        colorRow("Danger", PrismaSemanticColors.statusDangerDefault)
                        colorRow("Info", PrismaSemanticColors.statusInfoDefault)
                    }
                    group("Spacing") {
                        scalarRow("sp1", "\(Int(PrismaSpacing.sp1)) pt")
                        scalarRow("sp2", "\(Int(PrismaSpacing.sp2)) pt")
                        scalarRow("sp3", "\(Int(PrismaSpacing.sp3)) pt")
                        scalarRow("sp4", "\(Int(PrismaSpacing.sp4)) pt")
                        scalarRow("sp5", "\(Int(PrismaSpacing.sp5)) pt")
                        scalarRow("sp6", "\(Int(PrismaSpacing.sp6)) pt")
                        scalarRow("sp7", "\(Int(PrismaSpacing.sp7)) pt")
                        scalarRow("sp8", "\(Int(PrismaSpacing.sp8)) pt")
                    }
                    group("Radius") {
                        scalarRow("sm", "\(Int(PrismaRadius.sm)) pt")
                        scalarRow("md", "\(Int(PrismaRadius.md)) pt")
                        scalarRow("lg", "\(Int(PrismaRadius.lg)) pt")
                        scalarRow("full", "\(Int(PrismaRadius.full)) pt")
                    }
                    Spacer().frame(height: PrismaSpacing.sp7)
                }
                .padding(.horizontal, PrismaSpacing.sp5)
                .padding(.vertical, PrismaSpacing.sp5)
            }
        }
        .background(PrismaSemanticColors.surfaceBase.themed(scheme))
        .overlay(
            Rectangle()
                .frame(width: 1)
                .foregroundStyle(PrismaSemanticColors.borderSubtle.themed(scheme)),
            alignment: .leading
        )
    }

    @ViewBuilder
    private func group<Content: View>(_ label: String, @ViewBuilder content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp3) {
            Text(label.uppercased())
                .font(PrismaTypography.labelSm.font)
                .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
            VStack(alignment: .leading, spacing: PrismaSpacing.sp2) {
                content()
            }
        }
    }

    @ViewBuilder
    private func colorRow(_ label: String, _ token: PrismaSemanticColor) -> some View {
        let resolved = token.themed(scheme)
        let hex = hexString(of: resolved)
        HStack(spacing: PrismaSpacing.sp3) {
            RoundedRectangle(cornerRadius: 4)
                .fill(resolved)
                .frame(width: 20, height: 20)
                .overlay(
                    RoundedRectangle(cornerRadius: 4)
                        .strokeBorder(PrismaSemanticColors.borderSubtle.themed(scheme), lineWidth: 1)
                )
            Text(label)
                .font(PrismaTypography.bodySm.font)
                .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
                .frame(width: 140, alignment: .leading)
            Text(hex)
                .font(.system(.footnote, design: .monospaced))
                .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
        }
    }

    @ViewBuilder
    private func scalarRow(_ label: String, _ value: String) -> some View {
        HStack(spacing: PrismaSpacing.sp3) {
            Text(label)
                .font(PrismaTypography.bodySm.font)
                .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
                .frame(width: 80, alignment: .leading)
            Text(value)
                .font(.system(.footnote, design: .monospaced))
                .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
        }
    }
}

private func hexString(of color: Color) -> String {
    #if canImport(UIKit)
    var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
    UIColor(color).getRed(&r, green: &g, blue: &b, alpha: &a)
    return String(format: "#%02X%02X%02X", Int(r * 255), Int(g * 255), Int(b * 255))
    #else
    return ""
    #endif
}
